dev/ScrollViewer/ScrollViewer.cpp

```diff
@@ -483,11 +505,127 @@ void ScrollViewer::OnGotFocus(winrt::RoutedEventArgs const& args)
         m_focusInputDeviceKind == winrt::FocusInputDeviceKind::Mouse ||
         m_focusInputDeviceKind == winrt::FocusInputDeviceKind::Pen;
 
-    ShowIndicators();
+    UpdateScrollControllersAutoHiding();
+    if (AreScrollControllersAutoHiding())
+    {
+        UpdateVisualStates(true /*useTransitions*/, true /*showIndicators*/);
+    }
 }
 
 #pragma endregion
 
+bool ScrollViewer::AreScrollControllersAutoHiding()
+{
+    // Use the cached value unless it was invalidated.
+    if (s_autoHideScrollControllersValid)
+    {
+        return s_autoHideScrollControllers;
+    }
+
+    s_autoHideScrollControllersValid = true;
+
+    if (SharedHelpers::IsRS4OrHigher())
+    {
+        com_ptr<ScrollViewerTestHooks> globalTestHooks = ScrollViewerTestHooks::GetGlobalTestHooks();
+
+        if (globalTestHooks)
+        {
+            winrt::IReference<bool> autoHideScrollControllers = globalTestHooks->AutoHideScrollControllers();
+
+            if (autoHideScrollControllers)
+            {
+                // Test hook takes precedence over UISettings and registry key settings.
+                s_autoHideScrollControllers = autoHideScrollControllers.Value();
+                return s_autoHideScrollControllers;
+            }
+        }
+    }
+
+    if (s_uiSettings5)
+    {
+        // Use the 19H1+ UISettings property.
+        s_autoHideScrollControllers = s_uiSettings5.AutoHideScrollBars();
+    }
+    else if (SharedHelpers::IsRS4OrHigher())
+    {
+        // Use the RS4+ registry key HKEY_CURRENT_USER\Control Panel\Accessibility\DynamicScrollbars
+        s_autoHideScrollControllers = RegUtil::UseDynamicScrollbars();
+    }
+    else
+    {
+        // ScrollBars are auto-hiding prior to RS4.
+        s_autoHideScrollControllers = true;
+    }
+
+    return s_autoHideScrollControllers;
+}
+
+void ScrollViewer::OnAutoHideScrollBarsChanged(
+    winrt::UISettings const& uiSettings,
+    winrt::UISettingsAutoHideScrollBarsChangedEventArgs const& args)
+{
+    MUX_ASSERT(SharedHelpers::IsRS4OrHigher());
+
+    // OnAutoHideScrollBarsChanged is called on a non-UI thread, process notification on the UI thread using a dispatcher.
+    for (winrt::weak_ref<winrt::ScrollViewer> const& loadedScrollViewer : s_loadedScrollViewers)
```

Comment:
>s_loadedScrollViewers [](start = 74, length = 21)

It seems weird to me to keep track of this vector of loaded scroll viewers - why not just have each ScrollViewer subscribe to the event and dispatch back to itself, avoiding the need to manage this vector of weak refs?

Reply:
And I suspect this code has race conditions if multiple UI threads exist in the same process (ex: XAML islands on multiple threads, or secondary views in UWP), since this vector is not thread-safe.

---
In reply to: [285716351](https://github.com/microsoft/microsoft-ui-xaml/pull/731#discussion_r285716351) [](ancestors = 285716351)

Reply:
I was assuming a single UI thread, so I changed it back to what I initially had anyways: all ScrollViewer instances listen to the IUISettings5 changes. Thanks.